### ISSUE TITLE
Add DoRA (Weight-Decomposed Low-Rank Adaptation) Adapter Support

### ISSUE THREAD

**Issue by Dherya27** [axolotl-ai-cloud/axolotl#2723]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Hi Axolotl team, 
I'd like to contribute support for [DoRA (Weight-Decomposed Low-Rank Adaptation)](https://arxiv.org/abs/2402.09353) as an additional adapter type in Axolotl. DoRA extends LoRA by decomposing weight updates into magnitude and direction, providing more robust and efficient fine-tuning, especially at lower ranks.

#### Motivation:

1) DoRA is gaining traction in the community for its improved performance and robustness over LoRA.

2) Adding DoRA will give Axolotl users access to the latest parameter-efficient fine-tuning techniques.

#### References:

[DoRA: Weight-Decomposed Low-Rank Adaptation](https://arxiv.org/abs/2402.09353)

[Sebastian Raschka’s implementation and tutorial](https://magazine.sebastianraschka.com/p/lora-and-dora-from-scratch)

<img width="344" alt="Image" src="https://github.com/user-attachments/assets/463c81f6-4f6c-4da0-9e88-8dd25f7084b7" />
<img width="318" alt="Image" src="https://github.com/user-attachments/assets/9669499b-bdfd-4e2c-b29c-b18460f4bd08" />

<img width="344" alt="Image" src="https://github.com/user-attachments/assets/f2716d4c-cee1-46ab-940c-87423aa0364e" />
<img width="335" alt="Image" src="https://github.com/user-attachments/assets/b4554006-88ef-4e85-a6b9-e57a47633ba0" />

#### Proposed Implementation:

1) Add a new `dora.py` module implementing the DoRA logic as a PyTorch module.

2) Update `adapter.py` to support "dora" as an adapter type, similar to existing LoRA and QLoRA support.

3) Expose configuration options for DoRA rank and alpha in the YAML config.

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by xzuyn** [comment]
DoRA is already supported.

`peft_use_dora: true`

https://github.com/axolotl-ai-cloud/axolotl/pull/1363
https://github.com/axolotl-ai-cloud/axolotl/blob/5eb01f3df194ce6d663cebf2de8f3fb8fe7ec8e0/src/axolotl/utils/schemas/peft.py#L41

**Comment by NanoCode012** [comment]
Hello! Thank you @xzuyn for the quick answer :) 

Like said above, DoRA has been added a while back. We've added it to our config ref but understand that it's hard to find https://github.com/axolotl-ai-cloud/axolotl/blob/4a8af60d34d5761811ee3a4420bc7c8658ccf29a/docs/config.qmd#L394-L396

I could look into adding a PEFT doc for features we support later.
